ext/opcache/Optimizer/zend_inference.c

```diff
@@ -3285,7 +3285,15 @@ static int zend_update_type_info(const zend_op_array *op_array,
 			break;
 		case ZEND_FE_FETCH_R:
 		case ZEND_FE_FETCH_RW:
-			tmp = t2;
+			/* TODO: This can definitely be more precise than what I came up with. */
+			if (opline->op2_type == IS_CV && !(t2 & (MAY_BE_REF))) {
+				/* Infer that foreach ($array as $values) {...} is definitely not undefined within the loop body and doesn't have any of the old types. */
+				tmp = 0;
+			} else {
+				/* Handle cases such as `$i = &$obj->typedProperty`, which can change the type of assignments */
+				tmp = t2;
+			}
+			// fprintf(stderr, "Previously %x for FE_FETCH_ARR at line %d\n", (int)tmp, (int)opline->lineno);
```

Comment:
Please split this into a separate PR and remove the commented code.

I also don't really understand the particular logic you used here. Possibly what you want is just `tmp = (t2 & MAY_BE_REF)`? That's the only thing that should be inherited from the original variable...

Reply:
Done. 

> I also don't really understand the particular logic you used here. Possibly what you want is just `tmp = (t2 & MAY_BE_REF)`? That's the only thing that should be inherited from the original variable...

I wasn't familiar with why `t2` was being used or if there were edge cases that I wasn't aware of that would require `t2` (e.g. if this was deliberate for some op types). Apparently not.